### PR TITLE
Domino Model Compliance

### PR DESCRIPTION
**Pull request by coreyjadams**
# PhysicsNeMo Pull Request

## Description
This PR updates domino to comply with the latest model standards in physicsnemo v2.0.  Domino is now inheriting from physicsnemo.module, you can save / restore via physicsnemo tools, etc.  One issue that was tricky was domino uses the hydra configs directly in model construction.  These aren't JSON serializable, which is a problem.

The solution I built was to define default parameters to the model in a dictionary, which has extensions for attributes.  So, it's duck typing a dataclass / hydra config, essentially.  It allows conversion from hydra, and if a user passes an omegaconf to the model constructor it will quietly convert it to this dictionary type.  This ensure json `dumps` / `loads` works.

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -190,7 +190,8 @@ lint.fixable = ["I"]
 
 # Never enforce `E501` (line length violations),
 # and `S311` (random number generators)
-lint.ignore = ["E501", "S311"]
+# and `F722` which breaks jaxtyping annotations
+lint.ignore = ["E501", "S311", "F722"]
```
Comment (coreyjadams) [inline pyproject.toml:206]: This is to stop the jaxtyping / ruff warnings.  They are not real anyways.

Comment (peterdsharpe) [reply pyproject.toml:206]: Adding this link here for docs:
https://docs.kidger.site/jaxtyping/faq/#flake8-or-ruff-are-throwing-an-error

File: physicsnemo/models/domino/encodings.py [hunk 2]
```diff
@@ -14,36 +14,66 @@
...
 import torch.nn as nn
 from einops import rearrange
+from jaxtyping import Float
 
 from physicsnemo.nn import BQWarp
 
 from .mlps import LocalPointConv
 
 
 class LocalGeometryEncoding(nn.Module):
```
[the remark this answers is not in the document]
Comment (coreyjadams) [reply physicsnemo/models/domino/encodings.py]: Fixed.

File: physicsnemo/models/domino/encodings.py [hunk 3]
```diff
@@ -96,25 +165,57 @@ def forward(
...
             encoding_g_inner.append(geo_encoding_sampled)
+
+        # Concatenate and apply local point convolution
         encoding_g_inner = torch.cat(encoding_g_inner, dim=2)
         encoding_g_inner = self.local_point_conv(encoding_g_inner)
 
         return encoding_g_inner
 
 
 class MultiGeometryEncoding(nn.Module):
```
[the remark this answers is not in the document]
Comment (coreyjadams) [reply physicsnemo/models/domino/encodings.py]: Fixed

File: physicsnemo/models/domino/model.py [hunk 4]
```diff
@@ -156,21 +191,15 @@ def __init__(
         output_features_vol: int | None = None,
         output_features_surf: int | None = None,
         global_features: int = 2,
-        model_parameters=None,
+        model_parameters: Any = None,
     ):
-        """
-        Initialize the DoMINO model.
+        super().__init__(meta=ModelMetaData(name="DoMINO"))
```
[the remark this answers is not in the document]
Comment (coreyjadams) [reply physicsnemo/models/domino/model.py:190]: Declined.

File: physicsnemo/models/domino/config.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        for key, value in self.items():
+            if isinstance(value, dict) and not isinstance(value, Config):
+                self[key] = Config(value)
+
+    def __getattr__(self, key: str) -> Any:
+        """Allow attribute-style access to dict keys."""
+        try:
+            return self[key]
+        except KeyError:
+            raise AttributeError(f"'{type(self).__name__}' has no attribute '{key}'")
```
[the remark this answers is not in the document]
Comment (coreyjadams) [reply physicsnemo/models/domino/config.py:75]: I don't think this is necessary on hidden functions like this

**Comment by coreyjadams** [comment]
blossom-ci

**Comment by coreyjadams** [comment]
@greptile-apps can you review this PR again after changes?

File: physicsnemo/models/domino/mlps.py [hunk 6]
```diff
@@ -26,15 +27,61 @@
 
 
 class AggregationModel(Mlp):
```
[the remark this answers is not in the document]
Comment (coreyjadams) [reply physicsnemo/models/domino/mlps.py:29]: No this isn't necessary.

File: physicsnemo/models/domino/config.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+def _dataclass_to_dict(obj: Any) -> dict:
+    r"""Recursively convert a dataclass to a dict."""
+    result = {}
+    for field in fields(obj):
+        value = getattr(obj, field.name)
+        if is_dataclass(value) and not isinstance(value, type):
+            result[field.name] = _dataclass_to_dict(value)
+        else:
+            result[field.name] = value
+    return result
```
[the remark this answers is not in the document]
Comment (coreyjadams) [reply physicsnemo/models/domino/config.py:138]: As a private function I think we can skip this.

**Comment by coreyjadams** [comment]
Yes, the really nested config of the domino model + hydra compatibility + mandatory json serializability backed me into the corner of having to do this.  It's not ideal. I don't propose we do this for other models, either, this is really to be considered a stop-gap to bring domino to v2.0 standards without completely forcing users to re-write their hydra config files.
